### ISSUE TITLE
no motion mask learning

### ISSUE THREAD

**Issue by suvigy** [YihongSun/Dynamo-Depth#15]
Hi, 

Thanks for the great work, it is aligned how I'm thinking about the motion segmentation using unsupervised depth. I'm trying the concept for fisheye camera,  applying fisheye camera model. For the moment I only use inner crop just not to have to deal with the masks (of egocar, and out of radial fov regions). So actually it is now similar to the pinhole situation, but with different cameramodel. Thanks for the easy code understanding, it was easy to adapt the Backproject and Project3D functions to use different camera model. 
I also use a custom fisheye dataset. and don't use the loss for ground points. (I don't have infinite road issue when learning depth and pose)

The problem what I'm facing is at the motion mask learning. In the first phase it learns the depth and pose properly (no overfit to dynamic objects), in the second phase I'm not totally sure what to watch to judge if it learns properly. 

But in the motion mask learning, I can only see black images even when it is obvious that there is moving object. I had 30fps, so I've thought I reduce it to 7fps, but still black frames.  Somtimea there is motion mask for moving objects (mostly for side cams).
So in the last phase it will overfit to dynamic objects, and I have the same effect, if I had only depth and pose with static env conditions.

Questions:
 - Since I work with custom datasetI think there are some hyperparams that has to be changed. I tried the number of epochs, scheduler step size, num_steps_per_epoch for weight ramp. Maybe do you have **some hints which hyperparams are worth trying to change if I have issue with motion mask learning**? (i.e I haven't tried loss weights yet)
 - **Do you think fps matters** for motion mask learning?
 - In my dataset there are not really crowded scenes, so quite sparse on dynamic objects, **1-2 moving objects in the training windows**, is that change something? (In the paper you mentioned you used different numbers for weight ramping for each dataset)

**Comment by YihongSun** [comment]
Hello,

Thank you for your interest in our work!

The intuitions that you shared are inline with my experience. Motion mask learning is essential for mitigating the degenerate predictions for dynamic objects, however, the learning is tricky for the following reasons:

1. With a high sparsity penalization to prevent false positives / prevent modeling the entire scene as dynamic, there needs to be sufficient amount of dynamic objects to properly learn this mask prediction.
2. If there is not enough dynamic objects (positive examples), the simple learning objective here may fail due to heavy class imbalance. Often times, the failure solution would be to predict everything as static (black frames).

Connecting these points with you observation, my best guess is as follows:
1. Too high of fps would reduce the importance of motion mask since modeling dynamic objects as static ones would not incur a large reconstruction loss. This thus would encourage the motion mask to go to all zero. (Answer to Q2)
2. Too little dynamic objects in the dataset would make learning difficult due to the class imbalance. (Answer to Q3)

To mitigate this issue, my best guess would be to change the regularization for the motion mask, either reducing the strength of penalization or reformulating the loss function to further encourage positive prediction. (Answer to Q1)

Hope this helps!

**Comment by YihongSun** [comment]
Closing this issue for now. Please feel free to re-open it if you have other questions or concerns. Thanks!
